packages/relay/src/index.ts

```diff
@@ -116,48 +129,103 @@ class Relay implements IRelay {
       };
     }
 
-    return this.gossipSub.publish(this.pubSubTopic, msg);
+    return this.gossipSub.publish(pubSubTopic, msg);
   }
 
-  /**
-   * Add an observer and associated Decoder to process incoming messages on a given content topic.
-   *
-   * @returns Function to delete the observer
-   */
   public subscribe<T extends IDecodedMessage>(
     decoders: IDecoder<T> | IDecoder<T>[],
     callback: Callback<T>
   ): () => void {
-    const contentTopicToObservers = Array.isArray(decoders)
-      ? toObservers(decoders, callback)
-      : toObservers([decoders], callback);
-
-    for (const contentTopic of contentTopicToObservers.keys()) {
-      const currObservers = this.observers.get(contentTopic) || new Set();
-      const newObservers =
-        contentTopicToObservers.get(contentTopic) || new Set();
+    const pubSubTopicToContentObservers = this.createObservers(
+      decoders,
+      callback
+    );
 
-      this.observers.set(contentTopic, union(currObservers, newObservers));
-    }
+    this.addObservers(pubSubTopicToContentObservers);
 
     return () => {
-      for (const contentTopic of contentTopicToObservers.keys()) {
-        const currentObservers = this.observers.get(contentTopic) || new Set();
-        const observersToRemove =
-          contentTopicToObservers.get(contentTopic) || new Set();
+      this.removeObservers(pubSubTopicToContentObservers);
+    };
+  }
 
-        const nextObservers = leftMinusJoin(
-          currentObservers,
-          observersToRemove
+  private createObservers<T extends IDecodedMessage>(
+    decoders: IDecoder<T> | IDecoder<T>[],
+    callback: Callback<T>
+  ): Map<PubSubTopic, Map<ContentTopic, Set<Observer<T>>>> {
+    return Array.isArray(decoders)
+      ? toObservers(Array.from(this.pubSubTopics), decoders, callback)
+      : toObservers(Array.from(this.pubSubTopics), [decoders], callback);
+  }
+
+  private addObservers<T extends IDecodedMessage>(
+    pubSubTopicToContentObservers: Map<
+      PubSubTopic,
+      Map<ContentTopic, Set<Observer<T>>>
+    >
+  ): void {
+    for (const [
+      pubSubTopic,
+      contentTopicToObservers
+    ] of pubSubTopicToContentObservers.entries()) {
+      const existingContentObservers =
+        this.observers.get(pubSubTopic) ||
```

Comment:
```suggestion
        this.observers.get(pubSubTopic) ??
```

I believe I provided a reference for this.